In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

Imports the model on which the dataset have to be Trained 

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from catboost import CatBoostClassifier

Loading the dataset 

In [3]:
df = pd.read_csv(r'C:\Users\HP\PW skills\SELF placed\Machine Learning\projects\Mini Hackathon!\data\processed\processed_data.csv')

In [4]:
df.head()

Age  Gender  Country  self_employed  family_history  treatment  \
0   46       0        1              1               1          0   
1   29       0        1              0               1          1   
2   31       0        1              1               0          0   
3   46       0        1              0               0          1   
4   41       0        1              0               0          1   

   work_interfere  no_employees  remote_work  tech_company  ...  leave  \
0               1             0            1             1  ...      1   
1               1             3            0             1  ...      0   
2               0             0            1             1  ...      0   
3               1             2            1             1  ...      0   
4               0             5            0             0  ...      0   

   mental_health_consequence  phys_health_consequence  coworkers  supervisor  \
0                          0                        0          1           1   
1                          0                        0          1           0   
2                          0                        0          1           0   
3                          0                        0          1           1   
4                          0                        0          0           0   

   mental_health_interview  phys_health_interview  mental_vs_physical  \
0                        0                      1                   1   
1                        0                      0                   0   
2                        0                      0                   1   
3                        0                      0                   0   
4                        0                      1                   0   

   obs_consequence  Age_Group  
0                1    Group 4  
1                0    Group 2  
2                0    Group 2  
3                0    Group 4  
4                0    Group 4  

[5 rows x 25 columns]

spliting dataset into x and y

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 944 entries, 0 to 943
Data columns (total 25 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Age                        944 non-null    int64 
 1   Gender                     944 non-null    int64 
 2   Country                    944 non-null    int64 
 3   self_employed              944 non-null    int64 
 4   family_history             944 non-null    int64 
 5   treatment                  944 non-null    int64 
 6   work_interfere             944 non-null    int64 
 7   no_employees               944 non-null    int64 
 8   remote_work                944 non-null    int64 
 9   tech_company               944 non-null    int64 
 10  benefits                   944 non-null    int64 
 11  care_options               944 non-null    int64 
 12  wellness_program           944 non-null    int64 
 13  seek_help                  944 non-null    int64 
 14  anonymity 

In [6]:
X = df.drop(columns=['treatment','Age_Group'], axis=1) # features

y = df['treatment'] # target variable

In [7]:
X.shape, y.shape

((944, 23), (944,))

spliting the X,y in to X_train, X_test and y_train , y_test

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [9]:
models = {
    'RandomForest': RandomForestClassifier(),
    'XGBoost': XGBClassifier(),
    'DecisionTree': DecisionTreeClassifier(),
    'LogisticRegression': LogisticRegression(),
    'SVC': SVC(),
    'GaussianNB': GaussianNB(),
    'CatBoostClassifier': CatBoostClassifier()
}

creating the function through which all models are trained and predict the output

In [10]:
def train_and_evaluate(models, X_train, X_test, y_train, y_test):
    results = {}
    for model_name, model in models.items():
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)
        accuracy = accuracy_score(y_test, y_pred)
        results[model_name] = accuracy
        print(f"Model: {model_name}")
        print(f"Accuracy: {accuracy:.4f}")
        print("Classification Report:")
        print(classification_report(y_test, y_pred))
        print("Confusion Matrix:")
        print(confusion_matrix(y_test, y_pred))
        print("-" * 50)
    return results

In [11]:
train_and_evaluate(models, X_train, X_test, y_train, y_test)

Model: RandomForest
Accuracy: 0.7711
Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.51      0.63       108
           1       0.76      0.93      0.83       176

    accuracy                           0.77       284
   macro avg       0.79      0.72      0.73       284
weighted avg       0.78      0.77      0.76       284

Confusion Matrix:
[[ 55  53]
 [ 12 164]]
--------------------------------------------------
Model: XGBoost
Accuracy: 0.7289
Classification Report:
              precision    recall  f1-score   support

           0       0.70      0.51      0.59       108
           1       0.74      0.86      0.80       176

    accuracy                           0.73       284
   macro avg       0.72      0.69      0.69       284
weighted avg       0.72      0.73      0.72       284

Confusion Matrix:
[[ 55  53]
 [ 24 152]]
--------------------------------------------------
Model: DecisionTree
Accuracy: 0.6972
Classifica

{'RandomForest': 0.7711267605633803,
 'XGBoost': 0.7288732394366197,
 'DecisionTree': 0.6971830985915493,
 'LogisticRegression': 0.7746478873239436,
 'SVC': 0.6302816901408451,
 'GaussianNB': 0.7746478873239436,
 'CatBoostClassifier': 0.7535211267605634}

As per the result the LogisticRegression have the higest accuracy 

'GaussianNB': 0.7746478873239436,

'LogisticRegression': 0.7746478873239436,

'RandomForest': 0.7711267605633803,

'CatBoostClassifier': 0.7535211267605634

'XGBoost': 0.7288732394366197,

'DecisionTree': 0.6971830985915493,

'SVC': 0.6302816901408451,


which can be used for our machine learning

# Now performing the hyperparameter tuning 
for the best result from the Logistic Regression model and CatBoostClassifier

to find max accuracy model for machine 

using the GridSearchCV

In [16]:
from sklearn.model_selection import GridSearchCV

In [17]:
param_grid_lr = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],  # Inverse of regularization strength
    'penalty': ['l1', 'l2'],            # Norm used in the penalization
    'solver': ['liblinear', 'saga']      # Algorithm to use in the optimization problem
}

print("Parameter grid for Logistic Regression:")
display(param_grid_lr)

Parameter grid for Logistic Regression:


{'C': [0.001, 0.01, 0.1, 1, 10, 100],
 'penalty': ['l1', 'l2'],
 'solver': ['liblinear', 'saga']}

In [18]:
grid_search_lr = GridSearchCV(estimator=LogisticRegression(), param_grid=param_grid_lr,
                              scoring='accuracy', cv=5, n_jobs=-1, verbose=1)
grid_search_lr.fit(X_train, y_train)
best_lr = grid_search_lr.best_estimator_
y_pred_lr = best_lr.predict(X_test) 
accuracy_lr = accuracy_score(y_test, y_pred_lr)
print(f"Best Logistic Regression Model Accuracy: {accuracy_lr:.4f}")
print("Best Hyperparameters for Logistic Regression:")
print(grid_search_lr.best_params_)


Fitting 5 folds for each of 24 candidates, totalling 120 fits
Best Logistic Regression Model Accuracy: 0.7500
Best Hyperparameters for Logistic Regression:
{'C': 0.1, 'penalty': 'l1', 'solver': 'saga'}


By performing Hyperparameter tuning on the logistic Regression the 

Accuracy is now : 75 % 

the brest paremeters are : 'C': 0.1, 'penalty': 'l1', 'solver': 'saga'

Now performing the Hyperparameter tuning on the CatBoost classifier 

In [19]:
param_grid_catboost = {
    'iterations': [100, 250, 500],  # Number of boosting iterations
    'learning_rate': [0.01, 0.05, 0.1],  # Learning rate
    'depth': [4, 6, 8],  # Depth of the trees
    'l2_leaf_reg': [1, 3, 5, 7],  # L2 regularization term
    'border_count': [32, 64, 128],  # The number of splits for numerical features
    'verbose': [False] # Suppress verbose output during training
}

print("Parameter grid for CatBoost Classifier:")
display(param_grid_catboost)

Parameter grid for CatBoost Classifier:


{'iterations': [100, 250, 500],
 'learning_rate': [0.01, 0.05, 0.1],
 'depth': [4, 6, 8],
 'l2_leaf_reg': [1, 3, 5, 7],
 'border_count': [32, 64, 128],
 'verbose': [False]}

In [20]:
grid_search_cb = GridSearchCV(estimator=CatBoostClassifier(), param_grid=param_grid_catboost,
                                scoring='accuracy', cv=5, n_jobs=-1, verbose=1)
grid_search_cb.fit(X_train, y_train)
best_cb = grid_search_cb.best_estimator_
y_pred_cb = best_cb.predict(X_test)
accuracy_cb = accuracy_score(y_test, y_pred_cb)
print(f"Best CatBoost Classifier Model Accuracy: {accuracy_cb:.4f}")
print("Best Hyperparameters for CatBoost Classifier:")
print(grid_search_cb.best_params_)

Fitting 5 folds for each of 324 candidates, totalling 1620 fits
Best CatBoost Classifier Model Accuracy: 0.7535
Best Hyperparameters for CatBoost Classifier:
{'border_count': 32, 'depth': 6, 'iterations': 100, 'l2_leaf_reg': 1, 'learning_rate': 0.01, 'verbose': False}


for HyperParemeter tuning in the CatBoost Classifier 

the best accuracy is  : 75 which is the higest 

the params are : {'border_count': 32, 'depth': 6, 'iterations': 100, 'l2_leaf_reg': 1, 'learning_rate': 0.01, 'verbose': False}

In [ ]:
param_grid_nb = {
    'var_smoothing': np.logspace(0, -9, num=100)
}

# Initialize GridSearchCV
grid_search_nb = GridSearchCV(estimator=GaussianNB(), 
                              param_grid=param_grid_nb,
                              scoring='accuracy', 
                              cv=5, 
                              n_jobs=-1, 
                              verbose=1)

# Fit the grid search to your data
grid_search_nb.fit(X_train, y_train)

# Get the best estimator
best_nb = grid_search_nb.best_estimator_

# Make predictions and evaluate
y_pred_nb = best_nb.predict(X_test) 
accuracy_nb = accuracy_score(y_test, y_pred_nb)

print(f"Best GaussianNB Model Accuracy: {accuracy_nb:.4f}")
print("Best Hyperparameters for GaussianNB:")
print(grid_search_nb.best_params_)

Fitting 5 folds for each of 100 candidates, totalling 500 fits
Best GaussianNB Model Accuracy: 0.7606
Best Hyperparameters for GaussianNB:
{'var_smoothing': 0.001873817422860383}


Best GaussianNB Model Accuracy: 0.7606
Best Hyperparameters for GaussianNB:
{'var_smoothing': 0.001873817422860383}

In [21]:
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [10, 20, None],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2],
    'bootstrap': [True]
}


# --- Initialize GridSearchCV ---
grid_search_rf = GridSearchCV(estimator=RandomForestClassifier(), 
                              param_grid=param_grid_rf,
                              scoring='accuracy', 
                              cv=5, 
                              n_jobs=-1, 
                              verbose=1)

# --- Fit the grid search to your data ---
grid_search_rf.fit(X_train, y_train)

# --- Get the best estimator ---
best_rf = grid_search_rf.best_estimator_

# --- Make predictions and evaluate ---
y_pred_rf = best_rf.predict(X_test) 
accuracy_rf = accuracy_score(y_test, y_pred_rf)

print(f"Best RandomForest Model Accuracy: {accuracy_rf:.4f}")
print("Best Hyperparameters for RandomForest:")
print(grid_search_rf.best_params_)

Fitting 5 folds for each of 24 candidates, totalling 120 fits
Best RandomForest Model Accuracy: 0.7746
Best Hyperparameters for RandomForest:
{'bootstrap': True, 'max_depth': 20, 'min_samples_leaf': 2, 'min_samples_split': 5, 'n_estimators': 200}


We can take this model for future 

Best RandomForest Model Accuracy: **77.46**

Best Hyperparameters for RandomForest:

{'bootstrap': True, 'max_depth': 20, 'min_samples_leaf': 2, 'min_samples_split': 5, 'n_estimators': 200}